core/src/main/java/org/elasticsearch/search/aggregations/metrics/percentiles/ParsedPercentiles.java

```diff
@@ -0,0 +1,187 @@
+                return iterator.hasNext();
+            }
+
+            @Override
+            public Percentile next() {
+                Map.Entry<Double, Tuple<Double, String>> next = iterator.next();
+                return new InternalPercentile(next.getKey(), next.getValue().v1());
+            }
+        };
+    }
+
+    @Override
+    protected String getType() {
+        return type;
+    }
+
+    @Override
+    protected XContentBuilder doXContentBody(XContentBuilder builder, Params params) throws IOException {
+        if (keyed) {
+            builder.startObject(InternalAggregation.CommonFields.VALUES.getPreferredName());
+            for (Map.Entry<Double, Tuple<Double, String>> percentile : percentiles.entrySet()) {
+                builder.field(String.valueOf(percentile.getKey()), percentile.getValue().v1());
+
+                String valueAsString = percentile.getValue().v2();
+                if (valueAsString != null && valueAsString.isEmpty() == false) {
+                    builder.field(percentile.getKey() + "_as_string", percentile.getValue().v2());
+                }
+            }
+            builder.endObject();
+        } else {
+            builder.startArray(InternalAggregation.CommonFields.VALUES.getPreferredName());
+            for (Map.Entry<Double, Tuple<Double, String>> percentile : percentiles.entrySet()) {
+                builder.startObject();
+                {
+                    builder.field(InternalAggregation.CommonFields.KEY.getPreferredName(), percentile.getKey());
+                    builder.field(InternalAggregation.CommonFields.VALUE.getPreferredName(), percentile.getValue().v1());
+                    String valueAsString = percentile.getValue().v2();
+                    if (valueAsString != null && valueAsString.isEmpty() == false) {
+                        builder.field(InternalAggregation.CommonFields.VALUE_AS_STRING.getPreferredName(), valueAsString);
+                    }
+                }
+                builder.endObject();
+            }
+            builder.endArray();
+        }
+        return builder;
+    }
+
+    protected static void declarePercentilesFields(ObjectParser<? extends ParsedPercentiles, Void> objectParser) {
+        ParsedAggregation.declareCommonFields(objectParser);
+        objectParser.declareField((parser, percentiles, context) -> {
+            XContentParser.Token token = parser.currentToken();
+            if (token == XContentParser.Token.START_OBJECT) {
+                percentiles.keyed = true;
+                while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+                    if (token.isValue()) {
+                        if (token == XContentParser.Token.VALUE_NUMBER) {
+                            double key = Double.valueOf(parser.currentName());
+                            double value = parser.doubleValue();
+                            percentiles.percentiles.compute(key, (k, v) -> {
```

Comment:
Nice, didn't know `map.compute()`, seems to be handy. I still wonder if this logic could be a bit simpler to read if we had two maps instead of Map<Double, Tuple<Double, String>>.